File: frontend/services/dashboard.service.ts
import apiClient from './api';
import { CategoryStats } from './category.service';
import { TagStats } from './tag.service';

const ANALYTICS_URL = '/admin/analytics';

export interface ArticleTrend {
    date: string;
    count: number;
}

export interface AnalyticsData {
    total_articles: number;
    total_views: number;
    total_comments: number;
    total_categories: number;
    total_readers: number;
    article_trend: ArticleTrend[];
    top_categories: CategoryStats[];
    top_tags: TagStats[];
}

export interface AdminStats {
    total_articles: number;
    total_readers: number;
    total_categories: number;
    pending_posts: number;
    draft_posts: number;
    article_trend: number;
    reader_trend: number;
}

export interface AdminActivity {
    id: string;
    type: string;
    content: string;
    timestamp: string;
    user: string;
}

export interface DashboardAnalytics {
    date: string;
    articles: number;
    views: number;
}

export interface CategoryDistribution {
    name: string;
    value: number;
}

export interface DashboardData {
    stats: AdminStats;
    activities: AdminActivity[];
    analytics: DashboardAnalytics[];
    category_distribution: CategoryDistribution[];
}

export interface ArticleRelationStats {
    top_incoming_links: {
        article_id: string;
        article_title: string;
        incoming_count: number;
    }[];
    top_outgoing_links: {
        article_id: string;
        article_title: string;
        outgoing_count: number;
    }[];
    category_relations: {
        category_name: string;
        relation_count: number;
    }[];
    bidirectional_stats: {
        total_links: number;
        bidirectional_links: number;
        bidirectional_ratio: number;
    };
}

export interface AdvancedAnalyticsData {
    seo_stats: {
        missing_meta_description: number;
        missing_og_image: number;
        total_seo_records: number;
    };
    workflow_stats: {
        status: string;
        avg_days: number;
    }[];
    media_stats: {
        total_files: number;
        total_size: number;
    };
    versioning_stats: {
        avg_edits_per_article: number;
        total_versions: number;
    };
    content_graph: {
        total_relations: number;
        orphan_pages: number;
        avg_links_per_article: number;
    };
    content_health: {
        avg_tag_density: number;
        good_content_length: number;
    };
    relation_stats: ArticleRelationStats;
}

export const getDashboardData = async () => {
    const response = await apiClient.get<DashboardData>('/admin/dashboard');
    return response.data;
};

export const dashboardService = {
    async getAnalytics() {
        const response = await apiClient.get<AnalyticsData>(ANALYTICS_URL);
        return response.data;
    },

    async getAdvancedAnalytics() {
        const response = await apiClient.get<AdvancedAnalyticsData>('/admin/advanced-analytics');
        return response.data;
    },

    async getHierarchyStats() {
        const response = await apiClient.get<CategoryHierarchyStats>(`${ANALYTICS_URL}/hierarchy/stats`);
        return response.data;
    },

    async getCategoryTree() {
        const response = await apiClient.get<CategoryTreeData>(`${ANALYTICS_URL}/hierarchy/tree`);
        return response.data;
    },

    async getSuperDashboard() {
        const response = await apiClient.get<SuperDashboardData>('/admin/super-dashboard');
        return response.data;
    },

    async exportReport() {
        const response = await apiClient.get('/admin/export-dashboard', { responseType: 'blob' });
        return response.data;
    }
};

// Hierarchy-specific types
export interface CategoryHierarchyStats {
    root_count: number;
    child_count: number;
    avg_children: number;
    max_depth: number;
}

export interface CategoryNode {
    id: string;
    name: string;
    slug: string;
    article_count: number;
    level: number;
    children?: CategoryNode[];
}

export interface CategoryTreeData {
    roots: CategoryNode[];
}

// New Dashboard Types
export interface RoleCount {
    role_name: string;
    count: number;
}

export interface UserStats {
    total_users: number;
    active_users: number;
    inactive_users: number;
    role_distribution: RoleCount[];
}

export interface AuditLog {
    id: string;
    user_id: string;
    action: string;
    table_name: string;
    record_id: string;
    created_at: string;
}

export interface EngagementStats {
    total_comments: number;
    spam_comments: number;
    deleted_comments: number;
    total_shares: number;
    featured_count: number;
}

export interface SystemAnalytics {
    recent_audit_logs: AuditLog[];
    total_logs: number;
}

export interface SuperDashboardData {
    stats: AnalyticsData;
    advanced: AdvancedAnalyticsData;
    user_stats: UserStats;
    category_tree: CategoryNode[];
    category_hierarchy: CategoryHierarchyStats;
    engagement: EngagementStats;
    system: SystemAnalytics;
    pulse: SystemPulse;
    heatmap: ActivityHeatmap[];
    velocity: EditorialSpeed;
}

export interface SystemPulse {
    cpu_usage: number;
    memory_usage: number; // in bytes
    goroutines: number;
    db_connections: number;
    uptime_seconds: number;
    last_restart: string;
}

export interface ActivityHeatmap {
    day: string;
    hour: number;
    hits: number;
}

export interface EditorialSpeed {
    draft_to_publish_days: number;
    total_published: number;
}

File: backend/internal/service/dashboard_service.go
package service

import (
	"backend/internal/domain"
	"bytes"
	"encoding/csv"
	"fmt"
	"runtime"
)

type dashboardService struct {
	repo domain.DashboardRepository
}

func NewDashboardService(repo domain.DashboardRepository) domain.DashboardService {
	return &dashboardService{repo: repo}
}

func (s *dashboardService) GetAnalytics() (*domain.DashboardAnalyticsData, error) {
	return s.repo.GetAnalytics()
}

func (s *dashboardService) GetAdvancedAnalytics() (*domain.AdvancedAnalyticsData, error) {
	return s.repo.GetAdvancedAnalytics()
}

func (s *dashboardService) GetCategoryHierarchyStats() (*domain.CategoryHierarchyStats, error) {
	return s.repo.GetCategoryHierarchyStats()
}

func (s *dashboardService) GetCategoryTree() ([]domain.CategoryNode, error) {
	return s.repo.GetCategoryTree()
}

func (s *dashboardService) GetSuperDashboard() (*domain.SuperDashboardData, error) {
	data, err := s.repo.GetSuperDashboard()
	if err != nil {
		return nil, err
	}

	var m runtime.MemStats
	runtime.ReadMemStats(&m)

	data.Pulse.GoRoutines = runtime.NumGoroutine()
	data.Pulse.MemoryUsage = m.Alloc
	data.Pulse.CPUUsage = float64(runtime.NumCPU())

	return data, nil
}

func (s *dashboardService) ExportDashboardData() ([]byte, error) {
	data, err := s.GetSuperDashboard()
	if err != nil {
		return nil, err
	}

	b := &bytes.Buffer{}
	w := csv.NewWriter(b)

	// Section 1: General Stats
	w.Write([]string{"--- GENERAL STATISTICS ---"})
	w.Write([]string{"Metric", "Value"})
	w.Write([]string{"Total Articles", fmt.Sprintf("%d", data.Stats.TotalArticles)})
	w.Write([]string{"Total Views", fmt.Sprintf("%d", data.Stats.TotalViews)})
	w.Write([]string{"Total Comments", fmt.Sprintf("%d", data.Stats.TotalComments)})
	w.Write([]string{"Total Users", fmt.Sprintf("%d", data.UserStats.TotalUsers)})
	w.Write([]string{""})

	// Section 2: Engagement
	w.Write([]string{"--- ENGAGEMENT ---"})
	w.Write([]string{"Total Shares", fmt.Sprintf("%d", data.Engagement.TotalShares)})
	w.Write([]string{"Total Comments", fmt.Sprintf("%d", data.Engagement.TotalComments)})
	w.Write([]string{"Spam Comments", fmt.Sprintf("%d", data.Engagement.SpamComments)})
	w.Write([]string{"Featured Articles", fmt.Sprintf("%d", data.Engagement.FeaturedCount)})
	w.Write([]string{""})

	// Section 3: System Health
	w.Write([]string{"--- SYSTEM HEALTH ---"})
	w.Write([]string{"CPU Cores", fmt.Sprintf("%.0f", data.Pulse.CPUUsage)})
	w.Write([]string{"Goroutines", fmt.Sprintf("%d", data.Pulse.GoRoutines)})
	w.Write([]string{"DB Connections", fmt.Sprintf("%d", data.Pulse.DBConnections)})
	w.Write([]string{"Uptime (Seconds)", fmt.Sprintf("%d", data.Pulse.UptimeSeconds)})
	w.Write([]string{""})

	// Section 4: Editorial Velocity
	w.Write([]string{"--- EDITORIAL VELOCITY ---"})
	w.Write([]string{"Avg Days (Draft to Publish)", fmt.Sprintf("%.2f", data.Velocity.DraftToPublishDays)})
	w.Write([]string{"Total Published", fmt.Sprintf("%d", data.Velocity.TotalPublished)})

	w.Flush()
	return b.Bytes(), nil
}
